packages/app-store/ouraring/api/callback.ts

```diff
@@ -0,0 +1,72 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { WEBAPP_URL } from "@calcom/lib/constants";
+import { getSafeRedirectUrl } from "@calcom/lib/getSafeRedirectUrl";
+import prisma from "@calcom/prisma";
+
+import { decodeOAuthState } from "../../_utils/decodeOAuthState";
+import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import getInstalledAppPath from "../../_utils/getInstalledAppPath";
+
+const scopes = ["daily"];
+
+let client_id = "";
+let client_secret = "";
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { code } = req.query;
+  if (typeof code !== "string") {
+    res.status(400).json({ message: "No code returned" });
+    return;
+  }
+
+  const appKeys = await getAppKeysFromSlug("ouraring");
+  if (typeof appKeys.client_id === "string") client_id = appKeys.client_id;
+  if (typeof appKeys.client_secret === "string") client_secret = appKeys.client_secret;
+  if (!client_id) return res.status(400).json({ message: "Oura Ring client id missing." });
+  if (!client_secret) return res.status(400).json({ message: "Oura Ring client secret missing." });
+
+  const toUrlEncoded = (payload: Record<string, string>) =>
+    Object.keys(payload)
+      .map((key) => key + "=" + encodeURIComponent(payload[key]))
+      .join("&");
+
+  const body = toUrlEncoded({
+    client_id,
+    grant_type: "authorization_code",
+    code,
+    scope: scopes.join(" "),
+    redirect_uri: WEBAPP_URL + "/api/integrations/ouraring/callback",
+    client_secret,
+  });
+
+  const response = await fetch("https://cloud.ouraring.com/oauth/token", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
+    },
+    body,
+  });
+
+  const responseBody = await response.json();
+
+  if (!response.ok) {
+    return res.redirect("/apps/installed?error=" + JSON.stringify(responseBody));
+  }
+
+  responseBody.expires_in = Math.round(+new Date() / 1000 + responseBody.expires_in);
+
+  await prisma.credential.create({
+    data: {
+      type: "oura_ring_other_calendar",
+      key: responseBody,
+      userId: req.session?.user.id,
+      appId: "ouraring",
+    },
+  });
+  const state = decodeOAuthState(req);
+  return res.redirect(
+    getSafeRedirectUrl(state?.returnTo.replace("http:/", "http://")) ??
```

Comment:
For some reason returnTo has `http://` with a missing `/` so making sure it's there. Might need to do the same for `https`, will test in Cal-Staging (which is actually dev 😅 cc @keithwillcode)